apis/generators/v1alpha1/generator_types.go

```diff
@@ -14,8 +14,71 @@ limitations under the License.
 
 package v1alpha1
 
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
 type ControllerClassResource struct {
 	Spec struct {
 		ControllerClass string `json:"controller"`
 	} `json:"spec"`
 }
+
+type ClusterGeneratorSpec struct {
+	// +kubebuilder:pruning:PreserveUnknownFields
```

Comment:
Not a 100% convinced on this approach, but this makes it actually super easy to implement with minimal code change.

Reply:
Well, it works :D 

At a first glance i didn't like it because it's confusing (e.g. nested/repeated fields like `spec` or `kind`) and has potentially useless data (e.g. `metadata` in the nested structure). It could be easy to miss adding a new generator to the ClusterGenerator spec.

But on the other hand we can solve these issues.

First thoughts:

```yaml
apiVersion: "don'tknow"
kind: ClusterGenerator
spec:
  generator:
    kind: foobar
    spec:
      # whatever is in there
```

Maybe we don't need the `ClusterGenerator.spec.<thing>`. As we don't depend on the `<thing>` here. We can just call it `generator` (or anything else, really) and make it a `*apiextensions.JSON` type. 

Now i gotta run for :pizza:, will dig and play around with this later on. Looks very promising and VERY SIMPLE. I like!

Reply:
I know what you mean, yeah. I was depending on the `Kind` part to be as easy to leverage the existing way of figuring out that what kind of generator it is as possible. :D 

But yeah, I think with a bit of fiddling, having it as gen and then `*apiextensions.JSON` would most likely also work.

Reply:
So the reason that this is the full object is basically because of this: 

```go
	secretMap, err := gen.Generate(ctx, obj, r.Client, namespace)

```
Gets the whole spec, then parses it into the CRD of the respective implementation.

Now, some of those actually use the name. Some, only use the Spec. Either way, you'd have to define the entire object even if it is just `Generator`. This way, at least, the API _shows_ what's available. ¯\\\_(ツ)\_/¯

Reply:
Actually, that's not true. All of them only parse the spec. So the spec and the kind is enough I guess.

But instead of just `Generator` I kind of wanted something similar to SecretStore where you can autocomplete available settings. It's more user-friendly, than trying to pierce together what you are supposed to write there. 🤔

Reply:
Something like this maybe?
```go
type ClusterGeneratorSpec struct {
	Kind                      string                     `json:"kind"`
	ACRAccessTokenSpec        *ACRAccessTokenSpec        `json:"acrAccessTokenSpec,omitempty"`
	ECRAuthorizationTokenSpec *ECRAuthorizationTokenSpec `json:"ecrRAuthorizationTokenSpec,omitempty"`
	FakeSpec                  *FakeSpec                  `json:"fakeSpec,omitempty"`
	GCRAccessTokenSpec        *GCRAccessTokenSpec        `json:"gcrAccessTokenSpec,omitempty"`
	GithubAccessTokenSpec     *GithubAccessTokenSpec     `json:"githubAccessTokenSpec,omitempty"`
	PasswordSpec              *PasswordSpec              `json:"passwordSpec,omitempty"`
	STSSessionTokenSpec       *STSSessionTokenSpec       `json:"stsSessionTokenSpec,omitempty"`
	UUIDSpec                  *UUIDSpec                  `json:"uuidSpec,omitempty"`
	VaultDynamicSecretSpec    *VaultDynamicSecretSpec    `json:"vaultDynamicSecretSpec,omitempty"`
	WebhookSpec               *WebhookSpec               `json:"webhookSpec,omitempty"`
}
```

Reply:
I see and agree, UX is a valid concern, your proposal is better now.

Reply:
The original or the new one with just the spec plus a global kind in it? :)))

Reply:
the new one with just the spec plus a global kind in it. :)

Reply:
Nice. I'll do that ✌️👌😊

Reply:
Hum, this added to the complexity somewhat because now:
- I have to pass in the Kind to the map of the resource data ( that's what is used in GetGenerator to determine what generator is needed )
- I had to add an extra layer because we just loop through and find whatever is set first in the map and that could now be the `Kind` field. Maybe I could do a `if v == "kind" { continue }`... but it looks kinda nice now:

```yaml
apiVersion: generators.external-secrets.io/v1alpha1
kind: ClusterGenerator
metadata:
  name: my-generator
spec:
  kind: password
  generatorSpec:
    passwordSpec:
      length: 42
      digits: 5
      symbols: 5
      symbolCharacters: "-_$@"
      noUpper: false
      allowRepeat: true
```

Hum, maybe the double `spec` is weird?

Reply:
Maybe it's find, because in the previous one the Kind must be correct too. Also, in the previous one you didn't have to define the Name and such either i guess... :think

Reply:
Previously you could do this:
```yaml
apiVersion: generators.external-secrets.io/v1alpha1
kind: ClusterGenerator
metadata:
  name: my-generator
spec:
  password:
    kind: Password
    spec:
      length: 42
      digits: 5
      symbols: 5
      symbolCharacters: "-_$@"
      noUpper: false
      allowRepeat: true

```

So I guess it's still using the correct kind that you don't have to now. But also, the ignore unknown fields marker was ugly.

Reply:
I changed it to 

```yaml
apiVersion: generators.external-secrets.io/v1alpha1
kind: ClusterGenerator
metadata:
  name: my-generator
spec:
  kind: Password
  generator:
    passwordSpec:
      length: 42
      digits: 5
      symbols: 5
      symbolCharacters: "-_$@"
      noUpper: false
      allowRepeat: true
```

I think that's not that bad actually. And code isn't that bad either. :D

Reply:
may i ask why , you did not do this,
you i think you can already identify the `generatorReleated` based on `spec.kind` 

```yaml
apiVersion: generators.external-secrets.io/v1alpha1
kind: ClusterGenerator
metadata:
  name: my-generator
spec:
  kind: Password
  generator:
    #passwordSpec: 
    spec:
      length: 42
      digits: 5
      symbols: 5
      symbolCharacters: "-_$@"
      noUpper: false
      allowRepeat: true
````

Reply:
Because you need to define the object otherwise it doesn't know what it needs to translate it into. You can define all of the generators in there.

I can't embed all of them as they might have conflicting field names as well.